File: R/MEloc.R
# Copyright (C) 2009 
# Kim-Anh Lê Cao, ARC Centre of Excellence ins Bioinformatics, Institute for Molecular Bioscience, University of Queensland, Australia
# and Queensland Facility for Advanced Bioinformatics, The University of Queensland, Australia
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.



MEloc = function(
	jcross, 
	train, 
	test, 
	n, 
	nq, 
	nv, 
	ng, 
	indclass, 
	data.cat, 
	data.cont, 
	prop.kmeans, 
	means.kmeans, 
	var.kmeans,
	loc.ind){

# number of categories
numcat = vector(length = nq)
for(i in 1:nq){
	numcat[i] = nlevels(as.factor(data.cat[,i]))
}
nc = nlevels(as.factor(data.cat[,loc.ind]))    #number of cat in the location 

# -- read proportions and means from kmeans output and variance
prop = prop.kmeans
means.init = means.kmeans

var.init = array(0, dim = c(nv, nv, ng)) 

var.init = var.kmeans

# --- other initializations
lambda = array(0, dim = c(nq, max(numcat), ng))   # !!! I change the dim here
var.mat = array(0, dim = c(nv, nv, ng)) 
inv.mat = array(0, dim = c(nv, nv, ng)) 
det.val = vector(length = ng, mode = 'numeric')
mat.gum = matrix(nrow = n, ncol =ng)


main.res = main.loc(n, nv, nq, ng,nc, indclass, prop, numcat, var.init, inv.mat, det.val, prop.kmeans, means.init, data.cat, data.cont, train, test, lambda, mat.gum, loc.ind)

return(invisible(list(
prop = main.res$prop,
w = main.res$w,
loglik = main.res$vect.loglik,
mat.gum = main.res$mat.gum
)))
}


# ================= gating function and other functions for independence model =================

#------------------------norm function --------------------------------------------
norm.loc = function(x){
	 vect.norm = x/drop(sqrt(crossprod(x)))
	return(vect.norm)
}

norm2.loc = function(x){
	vect.norm = x/sum(x)
	return(vect.norm)
}

#------------------------init function --------------------------------------------

initmain.loc = function(nv, nq, ng, nc, numcat, var.init, inv.mat, det.val, means.init, lambda){

#----fill lambda
for(i in 1:nq){
	for(k in 1:ng){
		lambda[i,1:numcat[i],k] = rep(1/numcat[i], numcat[i])
	}
}

# -- variance and means
var.mat = var.init

mean = array(0, dim = c(nc, nv, ng))     # I change the dim here too  mean = array(0, dim = c(max(numcat), nv, ng))  
for( i in 1:nc){
	for(nvj in 1:nv){
	mean[i,nvj,1] =  means.init[1,nvj]   #+ runif(1) - 0.5
	mean[i,nvj,2] =  means.init[2,nvj]   #+ runif(1) - 0.5
}
}

#--inverse matrix and regularize if needed
if (det(var.mat[,,1]) <10^(-10)) {var.mat[,,1] = var.mat[,,1] + diag(0.2, nv, nv);inv.mat[,,1] = solve(var.mat[,,1])} else {inv.mat[,,1] = solve(var.mat[,,1])}
if (det(var.mat[,,2]) <10^(-10)) {var.mat[,,2] = var.mat[,,2] + diag(0.2, nv, nv);inv.mat[,,2] = solve(var.mat[,,2])}  else {inv.mat[,,2] = solve(var.mat[,,2])}

det.val[1] = det(solve(inv.mat[,,1]))
det.val[2] = det(solve(inv.mat[,,2]))


#-- initialize weights
w = matrix(nrow = ng, ncol = nv+nq)
for(k in 1:ng){
	w[k,] = 2 * runif(nv+nq) -1
}
w = t(apply(w,1, norm.loc))   #   !!!! norm w


return(invisible(list(
lambda = lambda,
var.mat = var.mat,
mean = mean,
inv.mat = inv.mat,
det.val = det.val,
w=w
)))

}

#============ gating function ==========================================================

main.loc = function(n, nv, nq, ng,nc, indclass, prop, numcat, var.init, inv.mat, det.val, prop.kmeans, means.init, data.cat, data.cont, train, test, lambda, mat.gum, loc.ind){

vect.loglik = vector(length = 10, mode = 'numeric')  #  number of iount
mat.gum = matrix(data = 0, nrow = n , ncol = 2)  # 

cat('test:', test, '\n')

# --- training set
data.cat.learn = as.matrix(data.cat[-test,] )      # remove test data
data.cont.learn = as.matrix(data.cont[-test,])

idt = data.cat.learn[,loc.ind]  #valeur de la location sans loo


init = initmain.loc(nv, nq, ng, nc, numcat, var.init, inv.mat, det.val, means.init, lambda)
lambda = init$lambda
var.mat = init$var.mat
mean = init$mean
inv.mat = init$inv.mat
det.val = init$det.val
w=init$w

# ============== EM ALGORITHM ====================
iount = 1 
while(iount <=10){                            #lim iount

#--- E-step
sw = vector(mode = 'numeric', length = ng)        # is initialized in E-step
swc = array(0, dim = c(ng, nq, max(numcat)))  
swxc = array(0, dim = c(ng, nv, nc))  
tau = matrix(data = 0 , nrow = length(train), ncol = ng)  # is initialized in the E-step
loglik = 0

mat.cat = matrix(data = 1 , nrow = length(train), ncol = ng)

for(j in 1:length(train)){
for(k in 1:ng){
for(iq in 1:nq){
	    mat.cat[j,k]=mat.cat[j,k]*lambda[iq,data.cat.learn[j,iq], k]
}#end iq
}}


#cat('estep', iount)
res.estep = .Fortran('eloc',	
		n = as.integer(length(train)),   ### careful here n = n-loo
		nv = as.integer(nv),
		ng = as.integer(ng),
		nc = as.integer(nc),
		nq = as.integer(nq),
		maxcat = as.integer(max(numcat)),  #
		x = as.matrix(data.cont.learn),
		xc = as.matrix(data.cat.learn),
		tau = as.matrix(tau),
		w = as.matrix(w),
		det = as.double(det.val),
		inv = as.array(inv.mat),
		indclass = as.integer(indclass),
		mean = as.array(mean),
		lambda = as.array(lambda),
		idt = as.integer(idt),
		t = as.double(prop),
		loglik = as.double(loglik),
		cat = as.matrix(mat.cat)
	)

tau = res.estep$tau
for(m in 1:length(train)){
	if(length(which(is.na(tau[m,])))==ng){tau[m,] = c(0.5, 0.5)}else{
	if(is.na(tau[m,1])) tau[m,1] = 1- tau[m,2]
	if(is.na(tau[m,1])) tau[m,2] = 1- tau[m,1]
	}
	if(sum(tau[m,])==0){tau[m,] = c(0.5, 0.5)}
}


sw = apply(tau, 2, sum, na.rm = TRUE)
if(any(sw==0)){sw[which(sw==0)] = 0.01}

for(k in 1:ng){
for(j in 1: length(train)){
for(l in 1:nv){
		swxc[k,l,idt[j]] = sum(swxc[k,l,idt[j]], tau[j,k] * data.cont.learn[j,l], na.rm=TRUE)
} # end l
	for(iq in 1:nq){
		swc[k,iq, data.cat.learn[j,iq]] = sum(swc[k,iq, data.cat.learn[j,iq]], tau[j,k], na.m=TRUE)
	}

}  #end j
} #end k

vect.loglik[iount] = res.estep$loglik

if(sum(is.na(prop)) !=0){
##prop = prop.ngme
prop = prop.kmeans
#next
cat('na in prop, estep', '\n')
next
}

# --- M-step
#      subroutine mstep(n,nv,ng,nc, nq, x,sw,swc,swxc,
#     +  tau, w, var, mean, lambda, loc.ind)

#cat('mstep', iount)
res.mstep = .Fortran('mloc',
		n = as.integer(length(train)),   ### careful here n = n-loo
		nv = as.integer(nv),
		ng = as.integer(ng),
		nc = as.integer(nc),
		nq = as.integer(nq),
		maxcat = as.integer(max(numcat)),
		x = as.matrix(data.cont.learn),
		xc = as.matrix(data.cat.learn),
		sw = as.vector(sw),
		swc = as.array(swc),
		swxc = as.array(swxc),
		tau = as.matrix(tau),
		w = as.matrix(w),
		var = as.array(var.mat),
		mean = as.array(mean),
		lambda = as.array(lambda),
		loc.ind = as.integer(loc.ind),
		numcat = as.vector(numcat),
		indclass = as.integer(indclass),
		t = as.double(prop)		
	)

lambda = res.mstep$lambda
for(k in 1:ng){
if(nq > 1){lambda[,,k] = t(apply(lambda[,,k],1,norm2.loc))} else{lambda[,,k] = norm2.loc(lambda[,,k])} 
}                                            #norm lambda


var.mat = res.mstep$var
mean = res.mstep$mean
w =  res.mstep$w
w = t(apply(w,1, norm.loc))   #   !!!! norm w
                  
prop = res.mstep$t

mean[which(mean=='Inf', arr.ind = TRUE)] = 0
mean[which(mean=='-Inf', arr.ind = TRUE)] = 0


#--inverse matrix and regularize if needed
if (det(var.mat[,,1]) <10^(-10)) {var.mat[,,1] = var.mat[,,1] + diag(0.2, nv, nv);inv.mat[,,1] = solve(var.mat[,,1])} else {inv.mat[,,1] = solve(var.mat[,,1])}
if (det(var.mat[,,2]) <10^(-10)) {var.mat[,,2] = var.mat[,,2] + diag(0.2, nv, nv);inv.mat[,,2] = solve(var.mat[,,2])}  else {inv.mat[,,2] = solve(var.mat[,,2])}

det.val[1] = det(solve(inv.mat[,,1]))
det.val[2] = det(solve(inv.mat[,,2]))

iount = iount+1
} # -- fin iount


#--------------------------------------subroutine check
#      subroutine check(n,nv,ng,nc, nq, ind,xcj,xj,
#     +  w, mean,lambda,inv,det,loc.ind,t,indclass,total)


#---- check for each sample in the test sample
indivtest=0
for(jtest in test){
#cat('jtest, check', jtest)
indivtest = indivtest+1

gum = vector(mode = 'numeric',length=ng)
res.check = .Fortran('checkloc',
		n = as.integer(n),   ### careful here n = n total = num
		nv = as.integer(nv),
		ng = as.integer(ng),
		nc = as.integer(nc),
		nq = as.integer(nq),
		ind = as.integer(jtest),    #the loo obs
		xcj = as.matrix(data.cat),
		xj = as.matrix(data.cont),
		w = as.matrix(w),
		mean = as.array(mean),
		lambda = as.array(lambda),
		inv = as.array(inv.mat),
		det = as.double(det.val),
		loc.ind = as.integer(loc.ind),
		t = as.double(prop),
		gum = as.double(gum),
		maxcat = as.integer(max(numcat))
	)

#predicted[jtest] = res.check$pred
#mat.gum[jtest,] = res.check$gum
mat.gum[test[indivtest],] = res.check$gum

}  # end jtest


return(invisible(list(
prop = prop,
w = w,
#predicted = predicted,
vect.loglik = vect.loglik,
mat.gum = mat.gum
)))
}  #end function main

File: src/meloc.f
C Copyright (C) 2009 
C Shu-Kay Ng, School of Medecine, Logan Campus, Griffith University Meadowbrook, Australia
C Kim-Anh Lê Cao, ARC Centre of Excellence ins Bioinformatics, Institute for Molecular Bioscience, University of Queensland, Australia
C and Queensland Facility for Advanced Bioinformatics, The University of Queensland, Australia
C
C This program is free software; you can redistribute it and/or
C modify it under the terms of the GNU General Public License
C as published by the Free Software Foundation; either version 2
C of the License, or (at your option) any later version.
C
C This program is distributed in the hope that it will be useful,
C but WITHOUT ANY WARRANTY; without even the implied warranty of
C MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C GNU General Public License for more details.
C
C You should have received a copy of the GNU General Public License
C along with this program; if not, write to the Free Software
C Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.





C fortran functions to be called by R
C location model + binary classification

c ------ SUBROUTINE E STEP ---------------------------------------
cKA: for R code removed iount, xlogl, u
cka removed PARAMETERS and common blocks
cKA added as arguments: det inv indclass mean nq lambda idt t xc
cka:  removed sw, swc, swxc
cka added maxcat, passe xc en double


c      subroutine estep(n,nv,ng,nc, nq, x, xc,sw,swc,swxc,
c     +  tau, w, det, inv, indclass, mean, lambda, idt, t,
c     +  loglik)

      subroutine eloc(n,nv,ng,nc, nq, maxcat, x, xc,
     +  tau, w, det, inv, indclass, mean, lambda, idt, t,
     +  loglik, cat)


      INTEGER n,nv,ng,nc, indclass, nq, maxcat
      double precision gum, loglik
      double precision twopi,al(ng),xc(n,nq)
      double precision x(n,nv),tau(n,ng)
      double precision u(n,ng), w(ng,nv+nq),sumga(ng)
cka      double precision cat(ng)
      double precision cat(n,ng)
      double precision mean(nc,nv,ng),t(ng),det(ng)
      double precision inv(nv,nv,ng)
      double precision lambda(nq,maxcat,ng)
      integer idt(n)
      intrinsic log, sqrt, exp



c   obtain the sufficient stat
      twopi=2.*3.141592653
      loglik = 0

c ----------LOOP on the samples without loo, here n = num-1
      do 800 j=1,n
        gum=0.

	do 400 k=1,ng

           sumga(k)=0.
	   u(j,k) = 0.

c	  cat(k)=1.
c	  do 230 i=1,nq
c	    cat(k)=cat(k)*lambda(i,xc(j,i), k)
c230       continue
c	print *, xc(j,1)
c	print *, x(j,1)

          al(k)=0.
	  do 300 l=1,nv
	    do 250 m=1,nv
	      al(k)=al(k)+(x(j,l)-mean(idt(j),l, k))*inv(l,m,k)*
     +	      (x(j,m)-mean(idt(j),m,k))
250         continue
300       continue

c KA : I increased the limit from 175 to 250
cc	  if (al(k).gt.200.) then
c	    write(30,*) 'al(k) gt 175, k', al(k), k
c	    print *, 'al(k) gt 175, k', al(k), k
cc	    al(k)=0.
cc	    goto 400
cc	  endif

	  al(k)=exp(-.5*al(k))/(sqrt(det(k))*(twopi)**(nv*.5))
ccc	  write(30,*) 'al(k)', al(k)

c KA I changed the limit from -30 to -40
cc350       if (t(k).lt.1.E-40.or.al(k).lt.1.E-40) then
c		write(30,*) 't(k).lt.1.E-40.or.al(k).lt.1.E-40', t(k), al(k)
c		print *, 't(k).lt.1.E-40.or.al(k).lt.1.E-40', t(k), al(k)
cc		goto 400
cc	  endif
	  sumga(k)=0.  

	  do 355 l=1,nq
	    sumga(k)=sumga(k)+w(k,l)*xc(j,l)
355       continue
	  do 360 l=1,nv
	    sumga(k)=sumga(k)+w(k,nq+l)*x(j,l)
360       continue
	  u(j,k)=1./(1.+exp(sumga(k)))

c KA for those who develop metastases, 
	  if (j.gt.indclass) u(j,k)=u(j,k)*exp(sumga(k))
	  gum=gum+t(k)*cat(j, k)*al(k)*u(j,k)
	  
400     continue

	if (gum.eq.0.) then
	  do 420 k=1,ng
	    tau(j,k)=0.
420       continue
c	  print *,'zero mixture density for entity in e-step',j
cc          do 200 k=1,ng
cc	  sw(k)=0.
cc	  do 80 i=1,nq
cc	  do 70 kk=1,nc
cc                swc(k,i,kk)=0.
cc70            continue
c80        continue
cc	  do 150 l=1,nv
cc	    do 90 kk=1,nc
cc	      swxc(k,l,kk)=0.
cc90          continue
cc150       continue
cc200       continue
	  goto 800
        endif

	do 550 k=1,ng
	  tau(j,k)=t(k)*cat(j,k)*al(k)*u(j,k)/gum
cc	  sw(k)=sw(k)+tau(j,k)
c KA: compute loglikelihood
	  loglik = loglik + tau(j,k)* log(t(k)*al(k)*u(j,k))
cc	  print *, 'loglik', loglik
c KA: added more than 2 cat in swc
cc	  do 430 i=1,nq
cc		swc(k,i,xc(j,i))=swc(k,i,xc(j,i))+tau(j,k)
cc430       continue

cc	  do 500 l=1,nv
cc	    swxc(k,l,idt(j))=swxc(k,l,idt(j))+tau(j,k)*x(j,l)
cc500       continue
550     continue
800   continue
      return
      end


C---------------------- M STEP -----------------------------
cKA for R code 
cka added maxcat

      subroutine mloc(n,nv,ng,nc, nq, maxcat, x, xc, sw,swc,swxc,
     +  tau, w, var, mean, lambda, lind, numcat, indclass, t)
c KA added ncat
      INTEGER ng,nc, ncat, indclass, maxcat
      double precision sw(ng), sum, w(ng,nv+nq)
      double precision swc(ng,nq,nc),swxc(ng,nv,nc)
      double precision x(n,nv)
      double precision mean(nc,nv,ng),t(ng)
      double precision var(nv,nv,ng)
      integer n,nv,gp
      double precision tau(n,ng)
c KA added numcat
      integer nq, numcat(nq)
      double precision lambda(nq,maxcat,ng)
      integer lind
c external function:
      external evalfw, HYBRD1
      integer sizehy,ifail, iflag,xc(n,nq)
c KA : be careful to allocate right dimension for work()
      double precision tem(nv+nq),xtol, taugp(n)
      double precision fvec(nv+nq)
      double precision work((nv+nq)*(3*(nv+nq) +13)/2 + 10)
      intrinsic exp

      ifail = 0

      sum=0.
      do 100 k=1,ng
	sum=sum+sw(k)
100   continue
c      print *, 'sum', sum

c  Update lambda  KA: added for more than 2 cat
      do 250 k=1,ng
	do 200 i=1,nq
	   ncat = numcat(i)
	   do 150 kk =1,ncat
		lambda(i,kk,k)=swc(k,i,kk)/sw(k)
150	   continue
200     continue
250   continue


c     Update t(k) and the mean and variance
      do 500 k=1,ng
	t(k)=sw(k)/sum
	do 400 l=1,nv
	  do 280 kk=1,nc
	    mean(kk,l,k)=swxc(k,l,kk)/swc(k,lind,kk)
280       continue
c	print *, 't(k)', t(k)
	  do 300 m=1,l
	    var(l,m,k)=0.
	  do 295 j=1,n 
	    do 290 kk=1,nc
	    var(l,m,k)=var(l,m,k)+(x(j,l)-mean(kk,l,k))
     +	      *(x(j,m)-mean(kk,m,k))*tau(j,k)
290         continue
295	continue
300       continue
400     continue
     	do 450 l=1,nv
	do 450 m=1,l
		var(l,m,k) = var(l,m,k)/sw(k)
		var(m,l,k)=var(l,m,k)
450   	continue
500   continue
cc      call matinv(nv,ng,var,inv,det)   to inverse later !!
c  Update w

      do 600 gp=1,ng
	do 510 l=1,nv+nq
	  tem(l)=w(gp,l)
510     continue

	do 520 j=1,n
	  taugp(j)=tau(j,gp)
520     continue

c ---- call HYBRD1
	xtol=0.000000001
c KA put a warning here, sizehy > N*(3*N +13)/2, N=nv+nq
cka	sizehy=2000
	sizehy=(nv+nq)*(3*(nv+nq) +13)/2 + 10
c	call HYBRD1(evalfw,nv+nq,tem,fvec,xtol,ifail,work,sizehy)

	call HYBRD1(nv+nq,tem,fvec,xtol,ifail,work,sizehy,
     +  n, x, xc, taugp, indclass, nv, nq)

c	if (ifail.ne.1) print *, 'ifail= ',ifail,gp
c	print *, (fvec(l),l=1,nv+nq)
	do 530 l=1,nv+nq
	  w(gp,l)=tem(l)
530     continue
600     continue
c       do 700 k=1,ng
c	write (30,*) 'swc: ', (swc(k,l,1),l=1,nq)
c	write (30,*) 'lambda: ', (lambda(k,l,1),l=1,nq)
c700   continue

      return 
      end

C ------------------------ EVALFW --------------------------------
C already defined somewhere else

c ---------------------- SUBROUTINE CHECK -----------------------      
c ka : removed indclass
cc      subroutine check(n,nv,ng,nc,ind,w)
cka removed pred, error

      subroutine checkloc(n,nv,ng,nc, nq, ind,xcj,xj,
     +  w, mean,lambda,inv,det,lind,t,
     +  gum, maxcat)
      INTEGER n,nv,ng,ind,out,nc, maxcat
      double precision gum(2),twopi,al(ng)
      double precision u(n,ng),w(ng,nv+nq),sumga(ng)
      double precision cat(ng)
      double precision mean(nc,nv, ng),t(ng),det(ng)
      double precision inv(nv,nv, ng)
      integer nq
      double precision lambda(nq,maxcat, ng)
      integer lind
      integer xcj(n,nv)
      double precision xj(n,nv)

      intrinsic log,sqrt

      twopi=2.*3.141592653
      j=ind
c here deals only with 2 classes
      do 800 jj=1,2 
        out=jj-1
        gum(jj)=0.

	do 400 k=1,ng
          cat(k)=1.
	  do 200 i=1,nq
	     cat(k)=cat(k)*lambda(i,xcj(j,i),k)
200       continue
cc	print *, 'cat', cat(k)
          al(k)=0.
	  do 300 l=1,nv
c ka : should be xj instead of x ?!
c ka and I changed idt(j) to xcj(j,lind)
	    do 250 m=1,nv
	      al(k)=al(k)+(xj(j,l)-mean(xcj(j,lind),l,k))*inv(l,m,k)*
     +	      (xj(j,m)-mean(xcj(j,lind),m,k))
250         continue
300       continue

c KA : changed here from 175 to 200
	  if (al(k).gt.200.) then
c	    write(30,*) 'al(k) gt 175, k', al(k), k
c	    print *, 'al(k) gt 175, k in check', al(k), k
	    al(k)=0.
	    goto 400
	  endif

	  al(k)=exp(-.5*al(k))/(sqrt(det(k))*(twopi)**(nv*.5))
350       if (t(k).lt.1.E-30.or.al(k).lt.1.E-30) then
c		write(30,*) 't(k).lt.1.E-40.or.al(k).lt.1.E-40', t(k), al(k)
c	        print *, 't(k)ltE-40.or.al(k)ltE-40 in check',t(k),al(k)		
		goto 400
	  endif
cc	print *, 'al(k)', al(k)
c ka: should be xcj and xj ?!
	  sumga(k)=0.  
	  do 355 l=1,nq
	    sumga(k)=sumga(k)+w(k,l)*xcj(j,l)
355       continue
	  do 360 l=1,nv
	    sumga(k)=sumga(k)+w(k,nq+l)*xj(j,l)
360       continue
cc	print *, 'sumga', sumga(k)

	  u(j,k)=1./(1.+exp(sumga(k)))
	  if (out.eq.1) u(j,k)=u(j,k)*exp(sumga(k))
	  gum(jj)=gum(jj)+t(k)*cat(k)*al(k)*u(j,k)
400     continue
cc	print *, 'gum', gum(jj)

	if (gum(jj).eq.0.) then
c	  print *,'zero mixture density for entity ',j
	  goto 800
        endif
800   continue

cka      pred=0
cka      if (gum(2).gt. gum(1)) pred=1
cka      error=0
c KA: indicate the class
c ka     if (j .le. indclass .and. pred .eq. 1) error=1
c ka     if (j .gt. indclass .and. pred .eq. 0) error=1
c      print *, j,gum(1),gum(2),pred,error
cc      print *, 'gum', gum(1),gum(2)
cc      total=total+error

cc      print *, 'total', total
      return 
      end

